### DISCUSSION TITLE
Meta issue: A quick design for the first release.

### DISCUSSION THREAD

**Discussion by He-Pin** [SymphonyQL/SymphonyQL#4] [Announcements]
Note: For now, I will type it in Chinese.

## 核心目标
1. Java first，Pekko-Stream  first， Scala first
2. 在第一版本中，交付核心的功能（比如query）。其他超集功能，以后再考虑。
3. Java first 中的注解支持、Scala first 中的宏支持，在上层模块中交付。

### 要点
1. 支持 query、mutation；stream 和 defer 功能可以后续交付
2. 因为是 Java first，所以核心的模型层要支持最后 手动builder，参考 graphql-java 这样的简单实现
3. 序列化使用 pekko 自带的jackson就好了 
4. 模型全部使用 pekko strema 的Source、Flow， 查询相当于是 Run the graph， 方便最后也支持 stream 和 defer。
5. 建议实现为基于预先分析的解释器，可以简化部分。

### 核心模块
1. parser; 这个使用pekko-http 的依赖 parboiled2，可以单独一个module，方便以后别个只用这个parser
2. 抽象一个 resolver，比如 MethodResolver, FieldResolver， resolver 的签名为 ctx, params => Source[T]
3. 抽象一个 action，比如 DBWriteAction之类的，mutation才需要，可以理解为 ctx, params => Flow[T, R, Effect]
4. schema 的定义参考官方的rfc 即可，scala 的 case class。支持值的转换，用户可能乱写。
5. anaylyzer 分析器，结合ctx 和 query对象，生成一个 preparedQuery/Mutation/Stream, 相当于RunnableGraph
6. 分析的过程中，埋入 trace，可以参考 zio的方式通过 using trace:Trace 搞
7. interpretor 解释器， 直接解释执行 preparedQuery/Mutation/Stream，相当于 RunnableGraph.run
8. object Symphony {} 是我们的入口类
9. web-console, 基于 pekko-http 提供embed console，可以直接访问和测试。
10. 在上面提供 javadsl、scaladsl。后续java 的外围搞基于 Spring 的，类似 Graphql-Spring， 支持基于注解来搞schema，或者手写。scala 的搞成宏或者手写。
11. 参考haxl， 在分析器下点功夫，解决 N+1 的问题

### 以后卷的方向
0. schema，从各种地方构建编排schema
1. 分析器，极致优化
2. 解释器，可以搞asm、bytebuddy之类的，越快越好。
3. resolver，外部扩展，比如提供 sql、http、grpc resolver
4. action ， 执行到db之类的
5. 自动的客户端~~

其他的：暂时就这么多吧，很累的。

**Comment by jxnu-liguobin** [comment]
params 都是类型化的T吧。
